man/fi_av.3.md

```diff
@@ -328,6 +348,16 @@ that calls to fi_av_insert following a call to fi_av_remove always reference a
 valid buffer in the fi_addr parameter.  Otherwise it may be difficult to
 determine what the next assigned index will be.
 
+If the address vector is configured with authorization keys, the fi_addr
```

Comment:
I think this behavior conflicts with FI_AV_USER_ID. We are planning to support FI_AV_USER_ID with FI_AV_AUTH_KEY.

Reply:
To handle this issue, maybe users need to provide the below context via the insert context arg.
```
struct fi_av_auth_key_context {
    fi_addr_t auth_key;
    void *context;
};
```
Thoughts?

Reply:
Ugh...  Separately, FI_AV_USER_ID and FI_AV_AUTH_KEY seem to work well with the existing API.

Zeroth Issue: You're about to experience my thought process, starting with me inserting this issue above the 5 that I wrote below first.

First issue: FI_AV_USER_ID is not identified as either a primary or secondary capability.  Primary means the app has to request it.  Secondary means the provider can offer it even if not requested.  This particular bit hits me as a secondary capability.  That leads to:

Second issue: On AV creation, the provider doesn't know if it needs to support FI_AV_USER_ID.  The API kind of makes a mess of this.  FI_AV_USER_ID is specified during insert.  Meaning some addresses may have a user ID, while others won't.  That's not a big deal, since the provider could internally set the 'user ID' to the actual fi_addr_t value.  However, it would be easier if the provider knew that the AV would store user IDs.  Goto second issue.

Third issue: We could make FI_AV_USER_ID a primary capability.  That would let the provider know if user IDs were going to be used.  No, that's not the issue.  The issue is that the shm provider doesn't even set the capability bit.  It simply accepts the flag.  That aligns closer with a use case of FI_AV_USER_ID being a secondary capability.  I don't know how cxi treats this flag.

Fourth issue: Given the above, what I want is: 1. Define FI_AV_USER_ID as a secondary capability.  2. Specify FI_AV_USER_ID as a flag on AV creation.  3. Keeping the flag on insert would be kept for compatibility.  The issue is that this changes the API, which we aren't quite compliant with anyway.

Fifth issue: Even with the above changes, the original problem being posed is still unresolved.  All I did was raise 5 more issues, though counting this issue as a new issue is questionable.

I didn't count the Zeroth issue in the Fifth issue because I added the Zeroth issue between writing the fifth issue and this statement.

Where was I?  Oh, yes, getting paid by the word.

I can think of some options.  I don't like asking the user to cast fi_addr_t * to some other structure.  So, I'm going to rule that possibility out.  We could have the user pass in an array of structs through the context parameter.  That's usable if we're okay not supporting asynchronous insertions.  I propose that feature for removal in 2.0 (#9009).  Second option is to have the user pass in twice the number of fi_addr_t values as the count.  Half would be the auth_key, and half the user ID.  Whether we split the array or alternate values doesn't matter much.  Finally, we could add a new call to set the user ID and do that separately.

Reply:
>That aligns closer with a use case of FI_AV_USER_ID being a secondary capability. I don't know how cxi treats this flag.

We are still working on adding support for FI_AV_USER_ID. I think we would treat this as a secondary cap flag. If with this cap flag set at the AV level, do we want users to still have the flexibility about which inserts FI_AV_USER_ID will apply to?

If it helps simplify the problem, could we limit each insert call to only be associated with a single auth_key?

Reply:
Defining FI_AV_USER_ID as a secondary capability doesn't work seamlessly.  That bit changes the behavior of the API, which the user has to agree with.  From the viewpoint of the user, I would consider introducing another flavor of capability bits -- optional bits.  These require the app to request, but may be cleared by the provider.  The main reason for adding this new flavor is so that app can avoid calling fi_getinfo multiple times toggling bits on/off in order to determine what features it can use.

The other proposed changes in issue 4 seem okay.  I'd rather configure the AV as either supporting user IDs or not.  Handling this per insertion call doesn't seem useful and just adds complexity.

As for the auth key, limiting insertion to a single key doesn't help.  Even if inserting a single address, the user needs to provide 2 fi_addr_t values.  I'm leaning toward adding a call to set/change the user ID for an inserted address.  There's a race between calling insert and setting the user ID where a completion could be reported.  But I think that race exists anyway, where a completion can be generated before the address has been inserted.

Reply:
>We could have the user pass in an array of structs through the context parameter. That's usable if we're okay not supporting asynchronous insertions.

As long as there is a `void * context` in the struct the user passes in, how does this limit async insertions?

Reply:
It's more that the API becomes difficult to describe and use.  The context param would need to point to something like:
```
struct blah blah {
    void *async_context;
    fi_addr_t *auth_key;  // or user ID
};
```
The size of auth_key array would come from the count param.  The auth_key could instead be auth_key[0], but that makes extending blah blah more difficult.

Reply:
Since the ABI is changing in this PR, I'll take a shot at adding in optional caps to better support FI_AV_USER_ID.

> I'm leaning toward adding a call to set/change the user ID for an inserted address.

If I followed correctly, this would result in an insert being 2 functions calls. Would this result in users have to introduce locks to achieve the expected behavior?

Reply:
Yes for 2 calls.  E.g. insert() + set_user_id().

I don't know that a lock is necessary.  Prior to the address being inserted, a message could be received from that peer.  The completion would indicate FI_ADDR_NOTAVAIL.  With 2 calls, the message could arrive between calling insert() and set_user_id().  However, from the viewpoint of the app, that's the same occurring before insert() being called.  That is, the app can't distinguish between the two situations.  So, until set_user_id() is set, the provider should use FI_ADDR_NOTAVAIL.

2 calls does allow for the app to change the user ID later, if desired, and incur the raciness of doing so.

Reply:
>2 calls does allow for the app to change the user ID later, if desired, and incur the raciness of doing so.

This is the raciness I was thinking about.

Reply:
I don't see there's any real race with the flow of insert + set_id.  Any race already exists prior to calling insert.  There's a race with calling set_id + set_id.  That's on the application to handle.  But I also don't expect apps to do that anyway.